File: library/xml-model/src/main/java/org/cru/godtools/xml/model/ImageGravity.java
package org.cru.godtools.xml.model;

import org.jetbrains.annotations.Contract;
import org.xmlpull.v1.XmlPullParser;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import timber.log.Timber;

public final class ImageGravity {
    static final int START = 1 << 0;
    public static final int END = 1 << 1;
    static final int TOP = 1 << 2;
    static final int BOTTOM = 1 << 3;
    private static final int CENTER_X = START | END;
    private static final int CENTER_Y = TOP | BOTTOM;
    static final int CENTER = CENTER_X | CENTER_Y;
    private static final int X_AXIS = START | END | CENTER_X;
    private static final int Y_AXIS = TOP | BOTTOM | CENTER_Y;

    @Nullable
    static Integer parse(@NonNull final XmlPullParser parser, @NonNull final String attribute,
                         @Nullable final Integer defValue) {
        return parse(parser.getAttributeValue(null, attribute), defValue);
    }

    @Nullable
    @Contract("_, !null -> !null")
    static Integer parse(@Nullable final String raw, @Nullable final Integer defValue) {
        if (raw != null) {
            try {
                int gravity = defValue != null ? defValue : CENTER;
                boolean seenX = false;
                boolean seenY = false;
                for (final String type : raw.split("\\s+")) {
                    switch (type) {
                        case "start":
                            if (seenX) {
                                throw new IllegalArgumentException("multiple X-Axis gravities in: " + raw);
                            }
                            gravity &= ~X_AXIS;
                            gravity |= START;
                            seenX = true;
                            break;
                        case "end":
                            if (seenX) {
                                throw new IllegalArgumentException("multiple X-Axis gravities in: " + raw);
                            }
                            gravity &= ~X_AXIS;
                            gravity |= END;
                            seenX = true;
                            break;
                        case "top":
                            if (seenY) {
                                throw new IllegalArgumentException("multiple Y-Axis gravities in: " + raw);
                            }
                            gravity &= ~Y_AXIS;
                            gravity |= TOP;
                            seenY = true;
                            break;
                        case "bottom":
                            if (seenY) {
                                throw new IllegalArgumentException("multiple Y-Axis gravities in: " + raw);
                            }
                            gravity &= ~Y_AXIS;
                            gravity |= BOTTOM;
                            seenY = true;
                            break;
                        case "center":
                            if (!seenX) {
                                gravity |= CENTER_X;
                            }
                            if (!seenY) {
                                gravity |= CENTER_Y;
                            }
                            break;
                        default:
                            // ignore unrecognized types
                    }
                }
                return gravity;
            } catch (final IllegalArgumentException e) {
                Timber.e(e, "error parsing ImageGravity");
            }
        }
        return defValue;
    }

    public static boolean isCenter(final int align) {
        return (align & CENTER) == CENTER;
    }

    /* X-Axis tests */

    public static boolean isCenterX(final int align) {
        return (align & X_AXIS) == CENTER_X;
    }

    public static boolean isStart(final int align) {
        return (align & X_AXIS) == START;
    }

    public static boolean isEnd(final int align) {
        return (align & X_AXIS) == END;
    }

    /* Y-Axis tests */

    public static boolean isCenterY(final int align) {
        return (align & Y_AXIS) == CENTER_Y;
    }

    public static boolean isTop(final int align) {
        return (align & Y_AXIS) == TOP;
    }

    public static boolean isBottom(final int align) {
        return (align & Y_AXIS) == BOTTOM;
    }
}

File: library/xml-model/src/main/java/org/cru/godtools/xml/model/ImageGravity.kt
package org.cru.godtools.xml.model

import org.xmlpull.v1.XmlPullParser
import timber.log.Timber

private const val BIT_START = 1 shl 0
private const val BIT_END = 1 shl 1
private const val BIT_TOP = 1 shl 2
private const val BIT_BOTTOM = 1 shl 3

private const val BIT_CENTER_X = BIT_START or BIT_END
private const val BIT_CENTER_Y = BIT_TOP or BIT_BOTTOM
private const val BIT_CENTER = BIT_CENTER_X or BIT_CENTER_Y

private const val MASK_X_AXIS = BIT_START or BIT_END or BIT_CENTER_X
private const val MASK_Y_AXIS = BIT_TOP or BIT_BOTTOM or BIT_CENTER_Y

inline class ImageGravity(internal val gravity: Int) {
    val isCenter get() = gravity and (MASK_X_AXIS or MASK_Y_AXIS) == BIT_CENTER
    val isCenterX get() = gravity and MASK_X_AXIS == BIT_CENTER_X
    val isCenterY get() = gravity and MASK_Y_AXIS == BIT_CENTER_Y
    val isStart get() = gravity and MASK_X_AXIS == BIT_START
    val isEnd get() = gravity and MASK_X_AXIS == BIT_END
    val isTop get() = gravity and MASK_Y_AXIS == BIT_TOP
    val isBottom get() = gravity and MASK_Y_AXIS == BIT_BOTTOM

    infix fun and(other: ImageGravity) = ImageGravity(gravity and other.gravity)
    infix fun or(other: ImageGravity) = ImageGravity(gravity or other.gravity)

    companion object {
        val START = ImageGravity(BIT_START)
        val END = ImageGravity(BIT_END)
        val TOP = ImageGravity(BIT_TOP)
        val BOTTOM = ImageGravity(BIT_BOTTOM)
        val CENTER = ImageGravity(BIT_CENTER)

        internal val NONE = ImageGravity(0)

        internal fun parse(raw: String?, defaultGravity: ImageGravity = CENTER): ImageGravity {
            if (raw == null) return defaultGravity

            try {
                var gravity = defaultGravity.gravity
                var seenX = false
                var seenY = false
                REGEX_SEQUENCE_SEPARATOR.split(raw).forEach {
                    when (it) {
                        "start" -> {
                            require(!seenX) { "multiple X-Axis gravities in: $raw" }
                            gravity = (gravity and MASK_X_AXIS.inv()) or BIT_START
                            seenX = true
                        }
                        "end" -> {
                            require(!seenX) { "multiple X-Axis gravities in: $raw" }
                            gravity = (gravity and MASK_X_AXIS.inv()) or BIT_END
                            seenX = true
                        }
                        "top" -> {
                            require(!seenY) { "multiple Y-Axis gravities in: $raw" }
                            gravity = (gravity and MASK_Y_AXIS.inv()) or BIT_TOP
                            seenY = true
                        }
                        "bottom" -> {
                            require(!seenY) { "multiple Y-Axis gravities in: $raw" }
                            gravity = (gravity and MASK_Y_AXIS.inv()) or BIT_BOTTOM
                            seenY = true
                        }
                        "center" -> {
                            if (!seenX) gravity = gravity or BIT_CENTER_X
                            if (!seenY) gravity = gravity or BIT_CENTER_Y
                        }
                    }
                }

                return ImageGravity(gravity)
            } catch (e: IllegalArgumentException) {
                Timber.tag("ImageGravity").e(e, "error parsing ImageGravity")
                return defaultGravity
            }
        }
    }
}

internal fun XmlPullParser?.getAttributeValueAsImageGravity(name: String, defaultGravity: ImageGravity) =
    ImageGravity.parse(this?.getAttributeValue(null, name), defaultGravity)
